### ISSUE TITLE
Support receiving structs as arguments for external functions

### ISSUE THREAD

**Issue by chriseth** [ethereum/solidity#1603]
- [x] modify the type system accordingly
- [x] write the decoder for memory
- [x] write the decoder for calldata

**Comment by axic** [comment]
@chriseth isn't this fully supported via ABIEncoderV2?

**Comment by chriseth** [comment]
This is still pending, mostly because the stack representation of such types is not yet defined.

**Comment by Amxx** [comment]
This issue is almost one year old, and is still blocking. It makes it impossible to write interfaces with structs as input. Any idea when we can expect a fix ?

**Comment by chriseth** [comment]
@Amxx you can use structs if the implementing function is public instead of external:

```
pragma solidity ^0.5.0;
pragma experimental ABIEncoderV2;

interface I {
    struct S { uint a; uint b; }
    function f(S calldata) external returns (uint);
}

contract C is I {
    function f(I.S memory _s) public returns (uint) {
        return _s.a;
    }
}
```

**Comment by chriseth** [comment]
Implemented in #6209
